src/elements/components/LoginProvider.tsx

```diff
@@ -0,0 +1,192 @@
+import React, { useCallback, useEffect, useRef, useState } from 'react';
+import { JSForm, Props as FormProps } from '../../Form';
+import { getStytchJwt } from '../../utils/browser';
+import {
+  defaultClient,
+  getAuthClient,
+  initInfo,
+  initState
+} from '../../utils/init';
+import { isAuthStytch } from '../../integrations/utils';
+import Spinner from './Spinner';
+import { isHrefFirebaseMagicLink } from '../../integrations/firebase';
+import LoaderContainer from './LoaderContainer';
+import { v4 as uuidv4 } from 'uuid';
+import { rerenderAllForms } from '../../utils/formHelperFunctions';
+/** TODO: These next 2 should maybe be dynamically imported, but having trouble with that
+ * combined 6.9k gzipped, so OK for now
+ */
+import { useIdleTimer } from 'react-idle-timer';
+import throttle from 'lodash.throttle';
+
+const TEN_SECONDS_IN_MILLISECONDS = 1000 * 10;
+const FIVE_MINUTES_IN_MILLISECONDS = 1000 * 60 * 5;
+
+export const authState = {
+  authClient: null,
+  authEmail: '',
+  authId: '',
+  authPhoneNumber: '',
+  // This is a flag so we only redirect to the login start step immediately
+  // after auth, not during other form navigation
+  redirectAfterLogin: false,
+  sentAuth: false,
+  setAuthId: (() => {}) as (newId: string) => void,
+  setFormCompleted: (() => {}) as (
+    formName: string,
+    completed: boolean
+  ) => void,
+  onLogin: () => {},
+  onLogout: () => {}
+};
+
+const LoginProvider = ({
+  authClient: authClientProp,
+  authId: authIdProp,
+  formProps,
+  loginPath,
+  onLogin = () => {},
+  onLogout = () => {},
+  children
+}: {
+  authClient?: any;
+  authId?: string;
+  formProps: FormProps;
+  loginPath?: string;
+  onLogin?: () => void;
+  onLogout?: () => void;
+  children: JSX.Element;
+}) => {
+  const [_internalId] = useState(uuidv4());
+  const formCompleted =
+    initInfo().sessions[formProps.formName]?.form_completed ?? false;
+
+  // Need to use this flag because when doing magic link login the onChange
+  // event doesn't seem to be added early enough to catch the first event which
+  // is the one containing the token. subsequent events do not contain the token
+  const hasAuthedRef = useRef(false);
+  // Use this render state to force re-evaluation of authId, since authState isn't reactive as-is
+  const [render, setRender] = useState(true);
+  const [showLoader, setShowLoader] = useState(false);
+
+  const logoutActions = () => {
+    hasAuthedRef.current = false;
+    authState.authId = '';
+    onLogout();
+  };
+
+  useEffect(() => {
+    if (
+      // We should set loader for new auth sessions
+      window.location.search.includes('stytch_token_type') ||
+      isHrefFirebaseMagicLink() ||
+      // and existing ones
+      getStytchJwt()
+    ) {
+      authState.redirectAfterLogin = true;
+      // We always need to set the redirect flag, but on local hosted forms we don't want to set the loader
```

Comment:
why don't we want to set the loader on localhost?

Reply:
I should expand the comment. The loader isn't set on localhost so that the loader doesn't appear on initial load for every form if you are also logged into the dashboard on localhost